.travis.yml

```diff
@@ -11,17 +11,23 @@ before_script:
   # Use Travis retry feature for farm tests since they are flaky
   - 'if [[ "$TOXENV" == "travis-test-farm"* ]]; then export TRAVIS_RETRY=travis_retry; fi'
   - export TOX_TESTENV_PASSENV=TRAVIS
+  - 'if [[ "$SNAP" == true ]]; then snap/local/build_and_install.sh; fi'
```

Comment:
It bothers me that this command is "simply" a part of the `before_script`, because:
- it is in fact the main purpose of this pipeline, and should benefit from a correct visibility in the pipeline for troubleshooting purpose
- it will run for each associated task, so one for the test task, and one for the snap task, while ideally, it should be an artifact that is reused on all tasks that needs it
- for multiarch build, this overhead will become very high
- failure to build the snap is a fail fast condition here: associated to the multiple call of snap builds described above, we will start several time the same broken build, wasting resources and time during the release and the nightly builds

In general I think this should be a dedicated stage, that would be before the `test` one. But this implies to be able to store pipeline artifacts, and Travis CI does not support that (it has only the capability to push artifacts to external storages, see https://docs.travis-ci.com/user/uploading-artifacts/). So I do not really have a solution here, but wanted to note that the situation is suboptimal, if you have an alternative in mind.

Reply:
I agree this is suboptimal and I think the best approach if we wanted to fix this would be to switch to a different CI provider (at least for the snap builds) as you mentioned in your overall review comment. We could also potentially look into using something like `s3` as shown at that link, but at that point, we're basically rolling our own version of features offered by other CI providers so I think I'd rather just switch.

If we wanted to switch, I think Azure Pipelines is really nice, but I find some of the authentication stuff like https://github.com/certbot/certbot/issues/7574 pretty frustrating for an open source project. GitHub Actions is potentially pretty promising with the downside of potentially slower builds and no tests for older versions of Windows. Circle CI is another option, however, I think we'd have to pay to get equivalent features as the first two options. That's not ideal, but we could maybe do that if once we pay, they offer the best product. I'm not sure if there are other options and I think switching requires more research/testing, but I'm personally leaning towards GitHub Actions right now.

With that said, I'm personally not that worried about this right now either. It's inelegant and inefficient, but I think it'll work, especially as an initial version of building our snaps.

Reply:
I am really hesitant to approve a code that you and I consider as not great, in particular considering the high level of quality generally expected in contributions to Certbot codebase. Indeed I consider that having a strong pipeline overcomes the problems with Mattermost interconnectivity, considering that it concerns very few people while the snap build is for users. Also Azure Pipelines contributors made good improvements in GitHub PR checks in the scope of open source contributions, as anyone can check the pipelines executed on Azure without a Microsoft account.

That is why I am still in favor in finding a better fitted tool, and Azure Pipeline has my preference. I also appreciate Circle CI and use it in several projects, but it needs a paid account to scale. GitHub Actions from my latest checks seems still quite beta, and is for now a subset (in term of functionalities and performances) of Azure Pipelines.

Anyway moving to another CI tool would be a long term solution and time consuming. Right now Travis and Azure Pipelines are already integrated in Certbot, so are well positionned to limit technological risks.

If you want, I can open on my own a draft PR for Azure Pipelines, and we can compare the two approaches.

Reply:
I'm surprised you think that the downsides of Travis are so great here that it is worth blocking making progress on snaps to rewrite our CI system. I say rewrite our CI system because the only purpose of stages as offered by Travis and used in this PR is to prevent the snap deployment if any of our other tests are failing. An equivalent approach in any other CI system such as Azure requires migrating all of our Travis tests to that system which is a significant undertaking and probably in all cases results in some downsides like we've discussed with the different CI providers.

What really are the downsides of Travis here?

> - it is in fact the main purpose of this pipeline, and should benefit from a correct visibility in the pipeline for troubleshooting purpose

In terms of the output shown by Travis, the only thing I can think of is the snap build output is folded when it passes which seems very minor to me. In terms of understanding this file, it also seems pretty minor. I could define the build differently here, but if you really think a change is required to unblock this work, I think the best approach would simply be to add a comment saying what `SNAP` does in the matrix below so it can be clearly and immediately understood.
> - it will run for each associated task, so one for the test task, and one for the snap task, while ideally, it should be an artifact that is reused on all tasks that need it
> - for multiarch build, this overhead will become very high
> - failure to build the snap is a fail fast condition here: associated to the multiple call of snap builds described above, we will start several time the same broken build, wasting resources and time during the release and the nightly builds

Unless I'm misunderstanding something, all of these concerns are about the potential downsides caused by future, unreviewed work. In this PR there is only one Travis job that defines `SNAP`. In all other jobs, this only adds a check of an environment variable.

If when we test it the overhead for multiarch builds which will mainly run while people aren't working on Certbot is unacceptable, maybe it makes sense to block that on rewriting our CI setup, but blocking us merging Robie's code into this repo on potential unacceptable inefficiencies in that work doesn't really make sense to me.

Out of all of the concerns you mentioned, unless I'm missing something, the only one that affects the state of things now is being able to understand what Travis does here which I don't think is that hard and seems like can be more easily resolved than abandoning Travis entirely.

I am in favor of careful design decisions, but I'm also in favor of small, incremental progress where possible and the phrase "the perfect is the enemy of the good". The purpose of this PR is to set up simple snap builds by mainly just merging Robie's code into this repo with minimal changes, unburdening him from maintaining it, and making it easier for us to continue to improve it as I don't think we can/should land many more changes in his repo. I think it accomplishes that purpose pretty well without any major current downsides.

Unless you can identify some significant current problems, I'd like to move forward with this PR as is and we can discuss rewriting our CI setup in the future. You working on that could be valuable, although as usual, the amount of time I'm willing to dedicate to that myself largely depends on if I can be convinced that it is the most important thing for us to be working on with our limited resources right now.

Reply:
My apologies, I think there is a misunderstanding because I did not interpretate correctly your intention in this PR. I was not thinking about migrating any tests from Travis to Azure, but just to write the specific logic of building, testing and deploying snaps into Azure instead of Travis because of the various reasons described in this PR comments.

But in my mind testing snaps was about launching integration tests only (so the `integration-external` tox target you added). Similarly to what was done in Robie repo originally.

So again in my request was concerning only the small incremental change done by this PR, which was then to "just" add a continuous delivery of snaps to the edge channel after a single test, and not imply to rethink right now the all CI system.

However I see know that your intention is to wait for all tests existing in Travis to succeed before pushing the snap to edge. I understand now more the reticence you have, and the trade your are doing between accepting a suboptimal Travis pipeline for snaps build versus a major refactoring of the CI system in Certbot.

Finally it comes all up to that question: is it really needed to wait for all tests to be run before pushing the snap? In my opinion, this is not necessary, and the `integration-external` is sufficient, for the following reasons:
- several tests (basically almost all from the extended tests suite) are not concerning snaps, either because they are testing a specific Python version, or testing Certbot installed bare on specific systems
- remaining tests that are testing the internal logic of Certbot are already executed continuously on master, either through PR integrations, or in nightly builds, giving us time to fix things before release time
- outcome of snap build would be for the edge channel, which I think is acceptable to do with limited blocking tests, in particular considering that tests that would fail at the time the snap is pushed and that we would not have the opportunity to see failing before would be from the extended test suite, which is not concerning the snap is I said before
- pushing the snap to beta is a manual action, giving the releaser the opportunity to review all outcomes of the release branch (including the extended test suite from Travis), and not give his sign off if there is any doubt

I sounds to me a quite agile workflow, with shorter multiple pipelines for different purposes even allowing "live" testing thanks to the edge channel if weird things are detected in the tests.

So it is now a trade-off between a suboptimal Travis pipeline for snaps versus not having the extended test suite blocking the push to edge channel because it is done in a dedicated pipeline in Travis. If you still think we really need to have all tests run before pushing to edge channel, I am totally agree with you that it does not worth to rethink the CI system right now. Otherwise, well I think it really makes sense to have the specific snap build/integration-test/deploy in Azure Pipeline.

What do you think of that? Sorry again for the verbosity of the discussion because we were not on the same page, because of a misunderstanding on my side.

Reply:
I do not think it is strictly necessary to block deploying the snap to edge on all tests passing, however, I think it's the behavior I'd ideally like to have as I can think of some (unlikely) scenarios where there are problems affecting the snap not caught by our tests on it. Maybe we found an incompatibility between Certbot and boulder. Maybe our unit tests or mypy found a problem in a code path not tested by the integration tests run on the snap. This is possible even if these tests are run on PRs because we do not require PRs to be up-to-date with `master` before being merged. Basically, if we've found unexpected problems in our code and we're unsure if they affect the snap, why deploy it?

The main downside I can think of in blocking deployments in this way limiting the discussion to the state of things in this PR is our full test suite [takes ~8 minutes longer](https://travis-ci.org/github/bmw/letsencrypt/jobs/678394225) because it's not run in parallel. I don't think this really matters at all for nightly or release builds, but it is kind of annoying for `test-` branches.

(The way things should ideally work in my mind is everything is test and built in parallel with deployments only happening on nightly/release builds after everything else has passed. This requires not only having better support for pipelines/stages than Travis has, but everything being in the same CI provider (or in different CI providers with very good integration with each other).)

Stepping back from that though, again my goal here was simply to merge Robie's code into this repo. I mainly used Travis because Robie used Travis so it made the migration slightly easier. I think almost all of your concerns about Travis are about interactions with future work built on this and are not caused by this PR alone. As far as I can tell, what we have here works just fine with no significant downsides, takes the code from Robie with minimal changes, and sets up snap builds nightly and for future releases.

If that's correct, I'd still personally like to move forward with this PR and save potential future problems for potential future work. If you want, I can remove stages entirely with this PR which eliminates the extra 8 minutes spent testing at the cost of potentially breaking things for people caused by problems caught by our other tests. I'm also happy to add a comment explaining what `SNAP` does when we set it for the snap job.

Reply:
No, for 8 min it does not worth it. We will talk about it in the future when the overhead will be more than 20 mins.

I close the discussion, for the sole purpose of what this PR wants to achieve, it does it.